src/plugins/plexil/blackboard_adapter.cpp

```diff
@@ -0,0 +1,646 @@
+ * @author Tim Niemueller
+ */
+
+/** Constructor.
+ * @param execInterface Reference to the parent AdapterExecInterface object.
+ */
+BlackboardPlexilAdapter::BlackboardPlexilAdapter(PLEXIL::AdapterExecInterface& execInterface)
+: InterfaceAdapter(execInterface),
+  BlackBoardInterfaceListener("PlexilBB")
+{
+}
+
+/** Constructor from configuration XML.
+ * @param execInterface Reference to the parent AdapterExecInterface object.
+ * @param xml A const reference to the XML element describing this adapter
+ * @note The instance maintains a shared pointer to the XML.
+ */
+BlackboardPlexilAdapter::BlackboardPlexilAdapter(PLEXIL::AdapterExecInterface& execInterface, 
+                                                 pugi::xml_node const xml)
+: InterfaceAdapter(execInterface, xml),
+  BlackBoardInterfaceListener("PlexilBB")
+{
+}
+
+/** Destructor. */
+BlackboardPlexilAdapter::~BlackboardPlexilAdapter()
+{
+}
+
+
+/** Initialize adapter.
+ * @return true if initialization was successful, false otherwise.
+ */
+bool
+BlackboardPlexilAdapter::initialize()
+{
+	logger_  = reinterpret_cast<fawkes::Logger *>(m_execInterface.getProperty("::Fawkes::Logger"));
+	blackboard_ =
+	  reinterpret_cast<fawkes::BlackBoard *>(m_execInterface.getProperty("::Fawkes::BlackBoard"));
+
+	namespace p = std::placeholders;
+	commands_ = {
+	  {"BB_open_for_reading",  std::bind(&BlackboardPlexilAdapter::bb_open_for_reading, this, p::_1)},
+	  {"BB_close",  std::bind(&BlackboardPlexilAdapter::bb_close, this, p::_1)},
+	  {"BB_read",  std::bind(&BlackboardPlexilAdapter::bb_read, this, p::_1)},
+	  {"BB_read_all",  std::bind(&BlackboardPlexilAdapter::bb_read_all, this, p::_1)},
+	  {"BB_print",  std::bind(&BlackboardPlexilAdapter::bb_print, this, p::_1)}
+	};
+
+	std::vector<std::string> lookups = { "BB_changed", "BB_int", "BB_real", "BB_bool",
+	                                     "BB_string", "BB_field_length",
+	                                     "BB_int_at", "BB_real_at", "BB_bool_at" };
+
+	for (const auto &c: commands_) {
+		PLEXIL::g_configuration->registerCommandInterface(c.first, this);
+	}
+
+	for (const auto &l: lookups) {
+		PLEXIL::g_configuration->registerLookupInterface(l, this);
+	}
+
+	blackboard_->register_listener(this, BlackBoard::BBIL_FLAG_DATA);
+
+	return true;
+}
+
+
+/** Start adapter.
+ * @return true if starting was successful, false otherwise.
+ */
+bool
+BlackboardPlexilAdapter::start()
+{
+	return true;
+}
+
+
+/** Stop adapter.
+ * @return true if successful, false otherwise.
+ */
+bool
+BlackboardPlexilAdapter::stop()
+{
+	return true;
+}
+
+
+/** Reset adapter.
+ * @return true if successful, false otherwise.
+ */
+bool
+BlackboardPlexilAdapter::reset()
+{
+	return true;
+}
+
+/** Shut adapter down.
+ * @return true if successful, false otherwise.
+ */
+bool
+BlackboardPlexilAdapter::shutdown()
+{
+	blackboard_->unregister_listener(this);
+
+	for (const auto& if_entry: ifs_read_) {
+		debugMsg("BlackboardAdapter:close",
```

Comment:
Where is `debugMsg` coming from? why not `logger_->log_debug`?

Reply:
debugMsg is plexil's debugging facility. It is used over log_debug to make it configurable similar to other plexil facilities (hence the entry in the config which is commented out by default). You can print all blackboard debug messages (using the default entry) or be more specific, e.g., to get only "close" messages you can configure "BlackboardAdapter:close" in the markers list.